src/replay_stage.rs

```diff
@@ -222,47 +215,111 @@ impl ReplayStage {
+                                Some(&vote_blob_sender),
+                                &ledger_entry_sender,
+                                &entry_height,
+                                &last_entry_id,
+                                entry_stream.as_mut(),
+                            ) {
+                                error!("{:?}", e);
+                            }
+                        } else {
+                            break;
+                        }
+
+                        let current_tick_height = bank.tick_height();
+
+                        // We've reached the end of a slot, reset our state and check
+                        // for leader rotation
+                        if max_tick_height_for_slot == current_tick_height {
+                            // Check for leader rotation
+                            let my_id = keypair.pubkey();
+                            let leader_id = Self::get_leader(&bank, &cluster_info);
+                            if leader_id != last_leader_id && my_id == leader_id {
+                                to_leader_sender
+                                    .send(TvuReturnType::LeaderRotation(
+                                        bank.tick_height(),
+                                        *entry_height.read().unwrap(),
+                                        *last_entry_id.read().unwrap(),
+                                    ))
+                                    .unwrap();
+                            }
+
+                            current_slot += 1;
+                            max_tick_height_for_slot = bank
+                                .leader_scheduler
+                                .read()
+                                .unwrap()
+                                .max_tick_height_for_slot(current_slot);
+                            last_leader_id = leader_id;
+                            continue;
+                        }
                     }
-                    last_leader_id = leader_id;
-
-                    match Self::process_entries(
-                        &bank,
-                        &cluster_info,
-                        &window_receiver,
-                        &keypair,
-                        vote_signer.as_ref(),
-                        Some(&vote_blob_sender),
-                        &ledger_entry_sender,
-                        &entry_height_.clone(),
-                        &last_entry_id.clone(),
-                        entry_stream.as_mut(),
-                    ) {
-                        Err(Error::RecvTimeoutError(RecvTimeoutError::Disconnected)) => break,
-                        Err(Error::RecvTimeoutError(RecvTimeoutError::Timeout)) => (),
-                        Err(e) => error!("{:?}", e),
-                        Ok(()) => (),
+
+                    // Block until there are updates again
+                    {
+                        let (cvar, lock) = &db_ledger.new_blobs_signal;
```

Comment:
I have some concerns about this pattern,  could you call db_ledger.wait_for_new_blobs(timeout) ?